File: packages/swift/MissionControlModels/Sources/MissionControlModels/Models/Task.swift
import Foundation

public enum TaskStatus: String, Codable {
    case pending
    case running
    case completed
    case failed
    case cancelled
}

public struct MCTask: Identifiable, Codable, Equatable {
    public let id: String
    public let requestId: String
    public let command: String
    public let status: TaskStatus
    public let createdAt: Date
    public let updatedAt: Date
    public var nodeId: String?
    public var result: TaskResult?

    public init(id: String, requestId: String, command: String, status: TaskStatus, createdAt: Date, updatedAt: Date, nodeId: String? = nil, result: TaskResult? = nil) {
        self.id = id
        self.requestId = requestId
        self.command = command
        self.status = status
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.nodeId = nodeId
        self.result = result
    }

    enum CodingKeys: String, CodingKey {
        case id = "taskId"
        case requestId, command, status, createdAt, updatedAt, nodeId, result
    }
}

public struct TaskResult: Codable, Equatable {
    public var exitCode: Int?
    public var stdout: String?
    public var stderr: String?
    public var errorMessage: String?

    public init(exitCode: Int? = nil, stdout: String? = nil, stderr: String? = nil, errorMessage: String? = nil) {
        self.exitCode = exitCode
        self.stdout = stdout
        self.stderr = stderr
        self.errorMessage = errorMessage
    }
}

File: packages/shared/src/schemas/task.ts
import { z } from 'zod';
import { LIMITS } from '../constants.js';

/**
 * Task status schema
 */
export const TaskStatusSchema = z.enum([
  'pending',
  'running',
  'completed',
  'failed',
  'cancelled',
]);

export type TaskStatus = z.infer<typeof TaskStatusSchema>;

/**
 * Task dispatch request schema (Hub -> Compute)
 */
export const TaskDispatchSchema = z.object({
  /** Unique task ID */
  taskId: z.string().uuid(),
  /** Request ID for audit trail */
  requestId: z.string(),
  /** Command to execute */
  command: z.string().min(1).max(LIMITS.MAX_INPUT_LENGTH),
  /** Working directory (within sandbox) */
  workingDir: z.string().optional(),
  /** Environment variables */
  env: z.record(z.string()).optional(),
  /** Timeout in milliseconds */
  timeoutMs: z
    .number()
    .min(1000)
    .max(LIMITS.MAX_TASK_TIMEOUT_MS)
    .default(LIMITS.DEFAULT_TASK_TIMEOUT_MS),
  /** Priority (higher = more urgent) */
  priority: z.number().min(0).max(100).default(50),
  /** Required capabilities (tags) for node selection */
  requiredCapabilities: z.array(z.string()).optional(),
});

export type TaskDispatch = z.infer<typeof TaskDispatchSchema>;

/**
 * Task result schema (Compute -> Hub)
 */
export const TaskResultSchema = z.object({
  /** Task ID */
  taskId: z.string().uuid(),
  /** Request ID */
  requestId: z.string(),
  /** Final status */
  status: TaskStatusSchema,
  /** Exit code (if completed) */
  exitCode: z.number().optional(),
  /** Standard output */
  stdout: z.string().max(LIMITS.MAX_OUTPUT_LENGTH).optional(),
  /** Standard error */
  stderr: z.string().max(LIMITS.MAX_OUTPUT_LENGTH).optional(),
  /** Error message (if failed) */
  errorMessage: z.string().optional(),
  /** Execution time in milliseconds */
  executionTimeMs: z.number().optional(),
  /** Node that executed the task */
  nodeId: z.string(),
});

export type TaskResult = z.infer<typeof TaskResultSchema>;

/**
 * Task query parameters
 */
export const TaskQuerySchema = z.object({
  status: TaskStatusSchema.optional(),
  nodeId: z.string().optional(),
  limit: z.number().min(1).max(100).default(20),
  cursor: z.string().optional(),
});

export type TaskQuery = z.infer<typeof TaskQuerySchema>;
